### PR TITLE
Metadata of C# (and other) Assemblies.

### PR DESCRIPTION
**Pull request by matthid**
Better support for loading C# (and other) dlls (we need IsFSharp to be **false** to get all methods in "MembersFunctionsAndValues"). This will be used by FSharp.Formatting for example.

I used this with success in https://github.com/tpetricek/FSharp.Formatting/pull/208 but I'm not exactly sure what I have done :)

If somebody can comment on this why it's needed or why it's wrong that would be nice. It looks like the right thing to do would be to get the ccu from the entitiy, but this seems to be impossible in this case (because it is not unresolved).

### CODE REVIEW

**Comment by matthid** [comment]
I have now added a C# project and a test-case showing the issue (I ensured that the test fails without the patch).
One sidenote: The API to analyse dll files is quite cumbersome to use and is basically a hack which first compiles a very simple F# project with the contents `module M`? I think it would be better to provide a more direct API in the future (to be used by FSharp.Formatting an other projects).

The commented lines in the test-method are not working (even with my patch), but they should eventually (imho).

I hope this can bring more attention to this issue!
PS: please tell me if that format is OK (ie. directly adding the C# project to the solution)

File: tests/service/CSharpProjectAnalysis.fs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        //String.IsNullOrWhiteSpace(members.[".ctor"].XmlDocSig) |> shouldEqual false
+        //String.IsNullOrWhiteSpace(members.["Method"].XmlDocSig) |> shouldEqual false
+        //String.IsNullOrWhiteSpace(members.["Property"].XmlDocSig) |> shouldEqual false
+        //String.IsNullOrWhiteSpace(members.["Event"].XmlDocSig) |> shouldEqual false
+        //String.IsNullOrWhiteSpace(members.["InterfaceMethod"].XmlDocSig) |> shouldEqual false
+        //String.IsNullOrWhiteSpace(members.["InterfaceProperty"].XmlDocSig) |> shouldEqual false
+        //String.IsNullOrWhiteSpace(members.["InterfaceEvent"].XmlDocSig) |> shouldEqual false
+
+        ()
+    | None -> ()
```
Comment (tpetricek) [inline tests/service/CSharpProjectAnalysis.fs]: I suppose that the code should fail in this case? (If the entity was not found, then something else went wrong...)

**Comment by tpetricek** [comment]
I don't know enough about F# compiler internals, but the test certainly looks correct to me!

Also +1 to providing simpler API for analyzing DLLs (so that the boilerplate code to create an empty project does not have to be repeated in F# Formatting and elsewhere).

File: tests/service/CSharpProjectAnalysis.fs (hunk 1 above)
Comment (matthid) [inline tests/service/CSharpProjectAnalysis.fs]: Correct :). While I noticed this shortly after pushing it, it's good to have a reminder comment, thanks. I guess it is no big deal for now (as that part is never executed), but I will fix that after getting an overall review/comment on the code.

**Comment by dungpa** [comment]
It seems wrong to include a solution test case into `FSharp.Compiler.Service.Tests` solution. Could you  move `CSharp_Analysis` solution to https://github.com/fsharp/FSharp.Compiler.Service/tree/master/tests/service/data instead?

**Comment by matthid** [comment]
The `CSharp_Analysis.sln` file was not really used anyway so I deleted it and moved the project to the destination you recommend.

Can the reference to the project and the project itself stay in the FSharp.Compiler.Service solution? Otherwise I'm not sure how to make the (new) test work by just opening the solution file and building FSharp.Compiler.Service.Tests...

**Comment by dungpa** [comment]
@matthid I think it's fine as is. You need CSharp_Analysis project to be compiled anyway.

In general, this is a nice improvement. I wonder what @dsyme thinks about this.

File: src/fsharp/vs/Symbols.fs [hunk 2]
```diff
@@ -194,7 +196,9 @@ and FSharpEntity(cenv:cenv, entity:EntityRef) =
     let isResolvedAndFSharp() = 
         match entity with
         | ERefNonLocal(NonLocalEntityRef(ccu, _)) -> not ccu.IsUnresolvedReference && ccu.IsFSharp
-        | _ -> true
+        | _ -> 
+            let ccu = defaultArg (ItemDescriptionsImpl.ccuOfItem cenv.g (getItem entity)) cenv.thisCcu 
+            ccu.IsFSharp
```
Comment (dsyme) [inline src/fsharp/vs/Symbols.fs]: I think these three lines should just be a single line that changes the "true" case on line 197 to "cenv.thisCcu.IsFSharp".  I think the getItem change is then no longer needed and the whole fix becomes one line.

File: src/fsharp/vs/Symbols.fs [hunk 3]
```diff
@@ -1788,7 +1792,7 @@ and FSharpAssembly internal (cenv, ccu: CcuThunk) =
     member __.FileName = ccu.FileName
     member __.SimpleName = ccu.AssemblyName 
     member __.IsProviderGenerated = ccu.IsProviderGenerated
-    member __.Contents = FSharpAssemblySignature(cenv,  ccu.Contents.ModuleOrNamespaceType)
+    member __.Contents = FSharpAssemblySignature((if ccu.IsUnresolvedReference then cenv else (new cenv(cenv.g, ccu, cenv.tcImports))),  ccu.Contents.ModuleOrNamespaceType)
```
Comment (dsyme) [inline src/fsharp/vs/Symbols.fs]: This line looks reasonable

**Comment by dsyme** [comment]
It looks good, I've made one suggestion for a code simplification

**Comment by matthid** [comment]
Thanks for looking into this, I pushed the simplification.

**Pull request merged**